### ISSUE TITLE
Periodically hang for a long time

### ISSUE THREAD

**Issue by joneswong** [ray-project/ray#2800]
### System information
- 32 cpu cores, 2 gpu
- **CentOS7**
- **source**
- **0.5.0**
- **py3.6**
- configuration:

```
prosthetics-apex-ddpg:
    env: ooxx
    run: APEX_DDPG
    stop:
        episode_reward_mean: 500
    checkpoint_freq: 20
    local_dir: /gruntdata/app_data/jones.wz/rl/nips2018/
    config:
        # problem setting
        clip_rewards: False
        horizon: 75

        # Misc
        timesteps_per_iteration: 2000
        num_workers: 20
        gpu: True
        min_iter_time_s: 30

        # Exploration
        schedule_max_timesteps: 1000000
        exploration_fraction: 0.1
        exploration_final_eps: 0.01
        noise_scale: 0.1
        per_worker_exploration: True
        worker_side_prioritization: False

        # Model
        actor_hiddens: [256, 128]
        actor_hidden_activation: selu
        critic_hiddens: [256, 128]
        critic_hidden_activation: selu

        # Optimization
        n_step: 3
        target_network_update_freq: 20000
        tau: 1.0
        use_huber: False
        grad_norm_clipping: 40.0
        train_batch_size: 256
        sample_batch_size: 32
        max_weight_sync_delay: 100
        learning_starts: 20000

        # Replay buffer
        buffer_size: 200000
```

### Describe the problem

I am using APEX_DDPG for an continuous control problem. The environment is widely used and wrapped by myself.
The experiment works well sometime, but will periodically hang for a while in some cases. It won't crash but there are not always tens of python process as what when it normally runs, e.g.,
![image](https://user-images.githubusercontent.com/6416895/44942676-20db5e00-adea-11e8-81f5-e0f79d4e5ac7.png)
![image](https://user-images.githubusercontent.com/6416895/44942686-36e91e80-adea-11e8-9043-12c3d2444d65.png)

Could you give me some suggestions to see what happened in those cases?

**Comment by ericl** [comment]
Hmm, I haven't seen hangs before, but maybe https://github.com/ray-project/ray/issues/1654 is relevant.

The thing to check would be if the learner thread is stalled, or if the agent is stalled on something else.

**Comment by joneswong** [comment]
I will add some stdout in agent to see which hangs. Do you have any idea according to this:
![image](https://user-images.githubusercontent.com/6416895/44943101-15406500-adf3-11e8-8aaa-523fff04fefc.png)

**Comment by ericl** [comment]
That looks pretty normal, so maybe it is a different issue.

**Comment by adoda** [comment]
I add some log, and found the training hang in the function:

![image](https://user-images.githubusercontent.com/35217964/47151920-bfc20800-d30d-11e8-8e7d-5644bdbab114.png)

```
obs, rew, done, info = self.envs[i].step(actions[i])
```
It may consume tens of minutes，and the function of "collect_episodes" in python/ray/rllib/evaluation/metrics.py will wait for all remote_evaluators finished.
So it will hang for tens of minutes.

**Comment by joneswong** [comment]
@ericl  Since Ray is not limited to stable simulators like gym, we'd better take these into consideration that some `remote_evaluator` have not finished their `sample_with_count()` task and thus the master hang at the metrics collecting step. I replaced the `ray.get()` command in `collect_episodes()` method by `ray.wait()` and @adoda  is testing this. If you think it is necessary, I will modify the metrics collecting step (may be with the help of `TaskPool`) and make a pr.

**Comment by ericl** [comment]
Hey Jones,

I think adding a wait with a timeout makes sense. We could print a warning
that some metrics were not collected in the iteration to make the situation
clear to the user. Also, using a task pool sounds good.

**Comment by simon-mo** [comment]
Seems to be fixed by #3096, if not please reopen
